redis/src/main/scala/zio/redis/Interpreter.scala

```diff
@@ -1,135 +1,236 @@
+      val makeBuffer = IO.effectTotal(ByteBuffer.allocateDirect(ResponseBufferSize))
+      ZLayer.fromEffect {
+        for {
+          address     <- getAddress
+          readBuffer  <- makeBuffer
+          writeBuffer <- makeBuffer
+        } yield new Connection(address, readBuffer, writeBuffer)
       }
     }
+
+    private final class Connection(
+      address: SocketAddress,
+      readBuffer: ByteBuffer,
+      writeBuffer: ByteBuffer
+    ) extends Service {
+
+      private def openChannel: Managed[IOException, AsynchronousSocketChannel] = {
+        val make = for {
+          channel <- IO.effect {
+                       val channel = AsynchronousSocketChannel.open()
+                       channel.setOption(StandardSocketOptions.SO_KEEPALIVE, Boolean.box(true))
+                       channel.setOption(StandardSocketOptions.TCP_NODELAY, Boolean.box(true))
+                       channel
+                     }
+          _       <- effectAsyncChannel[AsynchronousSocketChannel, Void](channel)(c => c.connect(address, null, _))
+        } yield channel
+        Managed.fromAutoCloseable(make).refineToOrDie[IOException]
+      }
+
+      override def connect: Managed[IOException, ReadWriteBytes] =
+        openChannel.map { channel =>
+          val readChunk: IO[Option[IOException], Chunk[Byte]] =
+            (for {
+              _     <- IO.effectTotal(readBuffer.clear())
+              _     <- effectAsyncChannel[AsynchronousByteChannel, Integer](channel)(c => c.read(readBuffer, null, _))
+                     .filterOrFail(_ >= 0)(new EOFException())
+              chunk <- IO.effectTotal {
+                         readBuffer.flip()
+                         val count = readBuffer.remaining()
+                         val array = Array.ofDim[Byte](count)
+                         readBuffer.get(array)
+                         Chunk.fromArray(array)
+                       }
+            } yield chunk).mapError {
+              case _: EOFException => None
+              case e: IOException  => Some(e)
+            }
+
+          def writeChunk(chunk: Chunk[Byte]): IO[IOException, Unit] =
+            IO.when(chunk.nonEmpty) {
+              IO.effectTotal {
+                writeBuffer.clear()
+                val (c, remainder) = chunk.splitAt(writeBuffer.capacity())
+                writeBuffer.put(c.toArray)
+                writeBuffer.flip()
+                remainder
+              }.flatMap { remainder =>
+                effectAsyncChannel[AsynchronousByteChannel, Integer](channel)(c => c.write(writeBuffer, null, _))
+                  .repeatWhileM(_ => IO.effectTotal(writeBuffer.hasRemaining))
+                  .zipRight(writeChunk(remainder))
+              }
+            }
+
+          new ReadWriteBytes {
+            override def read = Stream.repeatEffectChunkOption(readChunk)
+
+            override def write(chunk: Chunk[Byte]) = writeChunk(chunk)
+          }
```

Comment:
```suggestion
          new ReadWriteBytes {
            def read: Stream[IOException, Byte] = Stream.repeatEffectChunkOption(readChunk)

            def write(chunk: Chunk[Byte]): IO[IOException, Unit] = writeChunk(chunk)
          }
```

Looking at it again, I would inline `readChunk` and `writeChunk`.